File: minimongo/document.py
"""
Document validation and normalization helpers.
"""

from __future__ import annotations

import copy
import uuid

from .errors import DocumentValidationError


def normalize_document(doc: dict[str, object]) -> dict[str, object]:
    """
    Ensure the document is JSON-serializable-ish and has an _id.
    """
    if not isinstance(doc, dict):
        raise DocumentValidationError("document must be a dict")

    for key in doc.keys():
        if not isinstance(key, str):
            raise DocumentValidationError("document keys must be strings")

    normalized = copy.deepcopy(doc)
    normalized.setdefault("_id", str(uuid.uuid4()))
    return normalized

File: src/document.rs
use crate::{Document, WrongoDBError};
use crate::errors::DocumentValidationError;
use serde_json::Value;
use uuid::Uuid;

pub fn normalize_document(doc: &Document) -> Result<Document, WrongoDBError> {
    let mut normalized = doc.clone();

    let id_key = "_id".to_string();
    if !normalized.contains_key(&id_key) {
        normalized.insert(id_key, Value::String(Uuid::new_v4().to_string()));
    }

    // Keys are already String in Document; keep a lightweight validation for values:
    // ensure the document is JSON-serializable (Value always is).
    Ok(normalized)
}

pub fn validate_is_object(value: &Value) -> Result<(), WrongoDBError> {
    if !value.is_object() {
        return Err(DocumentValidationError("document must be a JSON object".into()).into());
    }
    Ok(())
}

